src/lib/reasoners/enum_rel.ml

```diff
@@ -28,32 +28,187 @@
+
+  let remove ~ex c d =
+    let constrs = HSS.remove c d.constrs in
+    let ex = Ex.union ex d.ex in
+    domain ~constrs ex
+end
+
+module Domains = struct
+  (** The type of simple domain maps. A domain map maps each representative
+      (semantic value, of type [X.r]) to its associated domain. *)
+  type t = {
+    domains : Domain.t MX.t;
+    (** Map from tracked representatives to their domain. *)
+
+    changed : SX.t;
+    (** Representatives whose domain has changed since the last flush
+        in [propagation]. *)
+  }
+
+  let pp ppf t =
+    Fmt.(iter_bindings ~sep:semi MX.iter
+           (box @@ pair ~sep:(any " ->@ ") X.print Domain.pp)
+         |> braces
+        )
+      ppf t.domains
+
+  let empty = { domains = MX.empty; changed = SX.empty }
+
+  let internal_update r nd t =
+    let domains = MX.add r nd t.domains in
+    let changed = SX.add r t.changed in
+    { domains; changed }
+
+  let get r t =
+    match Th.embed r with
+    | Cons (r, _) ->
+      (* For sake of efficiency, we don't look in the map if the
+         semantic value is a constructor. *)
+      Domain.singleton ~ex:Explanation.empty r
+    | _ ->
+      try MX.find r t.domains
+      with Not_found ->
+        Domain.unknown (X.type_info r)
+
+  let add r t =
+    match MX.find r t.domains with
+    | _ -> t
+    | exception Not_found ->
+      (* We have to add a default domain if the key `r` isn't in map in order to
+         be sure that the case-split mechanism will attempt to choose a value
+         for it. *)
+      let nd =
+        match Th.embed r with
+        | Cons (r, _) ->
+          Domain.singleton ~ex:Explanation.empty r
+        | _ ->
+          Domain.unknown (X.type_info r)
+      in
+      internal_update r nd t
+
+  (** [update r d t] replaces the domain of [r] in [t] by [d]. The
+      representative [r] is marked [changed] after this call if the domain
+      [d] isn't equal to the old one. *)
+  let update r d t =
+    let od = get r t in
+    (* For sake of completeness, the domain [d] has to be a subset
+       of the old domain of [r]. *)
+    assert (not (Options.get_enable_assertions ()) || Domain.subset d od);
+    if Domain.equal od d then
+      t
+    else
+      internal_update r d t
+
+  let remove r t =
+    let domains = MX.remove r t.domains in
+    let changed = SX.remove r t.changed in
+    { domains ; changed }
+
+  (** [subst ~ex p v d] replaces all the instances of [p] with [v] in all
+      domains, merging the corresponding domains as appropriate.
+
+      The explanation [ex] justifies the equality [p = v].
+
+      @raise Domain.Inconsistent if this causes any domain in [d] to become
+             empty. *)
+  let subst ~ex r nr t =
+    match MX.find r t.domains with
+    | d ->
+      let nnd =
+        match MX.find nr t.domains with
+        | nd -> Domain.intersect ~ex d nd
+        | exception Not_found -> d
+      in
+      let t = remove r t in
+      internal_update nr nnd t
```

Comment:
`update` should be used rather than `internal_update` to avoid adding unnecessary explanations (and also avoid storing domains for constants)

Reply:
If we don't store at all domains for the constructor, we should use `get` in `subst` as follows to get the appropriate 
intersection:
```ocaml
  let subst ~ex r nr t =
    match MX.find r t.domains with
    | d ->
      let nd = Domain.intersect ~ex d (get nr t) in
      let t = remove r t in
      update nr nd t

    | exception Not_found -> t
```

Reply:
Ah right! In fact this:

```ocaml
      let nnd =
        match MX.find nr t.domains with
        | nd -> Domain.intersect ~ex d nd
        | exception Not_found -> d
      in
```

was a soundness bug because `update` does not intersect and so we lose the explanation `ex` if `nr` is not in `t.domains`.

Reply:
I also rename `update` into `tighten`. `update` is too generic.